new ban validator functionality

Vitomir2 opened the pull request:
- create a new contract and implement a ban function for the validators that will be banned;
- create a new internal function _withdraw to handle the withdraw process and re-use;
- write the tests for the new inspector contract;
- fix some other tests;

R-Santev commented on contracts/ValidatorSet/modules/Inspector/Inspector.sol:
```diff
@@ -0,0 +1,62 @@
+
+    function _burnPenalty(address validator) private {
+        _burn(validator, validatorPenalty);
+        LiquidStaking._collectTokens(validator, validatorPenalty);
```
collect tokens must work differently. Will give you more information further.

R-Santev commented on test/ValidatorSet/Delegation.test.ts:
```diff
@@ -762,7 +762,7 @@ export function RunDelegationTests(): void {
         const maxTopUps = 52; // one cannot top-up more than 52 times
         for (let i = 0; i < maxTopUps; i++) {
-          const delegatingAmount = this.minDelegation.mul(i + 1).div(5);
+          const delegatingAmount = this.minDelegation.div(i + 1); //  mul(i + 1).div(5);
```
Why is this changed and a comment is added?

Vitomir2 replied:
Forgot this, but will revert some tet changes.

R-Santev commented on test/ValidatorSet/Inspector.test.ts:
```diff
@@ -0,0 +1,82 @@
+
+export function RunInspectorTests(): void {
+  describe("Ban a validator", function () {
+    it("should revert if try to ban zero address", async function () {
```
Why banning the zero address is a problem in case it has some balance? I don't think we need this check in practise.

Vitomir2 replied:
Well, I just thought there is no way the zero address to become a validator, but yeah, even without this check, the functionality will work fine.

R-Santev commented on test/ValidatorSet/Inspector.test.ts:
```diff
@@ -0,0 +1,82 @@
+      const penalizedValidator = this.signers.validators[0];
+      const stakedAmount = await validatorSet.balanceOf(penalizedValidator.address);
+
+      await validatorSet.connect(this.signers.governance).banValidator(this.signers.validators[0].address);
```
you can use the `penalizedValidator` variable here.

R-Santev commented on test/ValidatorSet/Inspector.test.ts:
```diff
@@ -0,0 +1,82 @@
+      expect(await validatorSet.balanceOf(penalizedValidator.address), "balanceOf").to.be.equal(0);
+      expect(await validatorSet.pendingWithdrawals(penalizedValidator.address), "withdrawals").to.be.equal(
+        stakedAmount.sub(validatorBanPenalty)
+      );
```
A check for some of the state changes that must occur is not made (e.g. stateSync)

R-Santev commented on test/ValidatorSet/Inspector.test.ts:
```diff
@@ -0,0 +1,82 @@
+      await registerValidator(validatorSet, this.signers.governance, newValidator);
+
+      const stakerValidatorSet = validatorSet.connect(newValidator);
+      await stakerValidatorSet.stakeWithVesting(VESTING_DURATION_WEEKS, {
```
Maybe you were able to use a fixture with a preset vesting position?

R-Santev commented on test/ValidatorSet/Inspector.test.ts:
```diff
@@ -0,0 +1,82 @@
+      const amountAfterBanPenalty = this.minStake.sub(validatorBanPenalty);
+      // hardcode the penalty percent by 0.3% a week
+      const bps = (VESTING_DURATION_WEEKS - 1) * 30;
+      const unstakePenalty = amountAfterBanPenalty.mul(bps).div(DENOMINATOR);
```
The process must be unstake -> unstake vested position penalty applied -> apply an additional penalty for the ban

R-Santev commented on test/ValidatorSet/Staking.test.ts:
```diff
@@ -72,7 +72,7 @@ export function RunStakingTests(): void {
       const { validatorSet } = await loadFixture(this.fixtures.stakedValidatorsStateFixture);
 
-      const unstakeInsufficientAmount = this.minStake.add(hre.ethers.utils.parseEther("5"));
+      const unstakeInsufficientAmount = this.minStake.mul(3);
```
Why is this changed?

R-Santev commented on test/ValidatorSet/Staking.test.ts:
```diff
@@ -105,8 +105,8 @@ export function RunStakingTests(): void {
 
-      const unstakeInsufficientAmount = hre.ethers.utils.parseEther("2.5");
-      const stakeTooLowAmount = hre.ethers.utils.parseEther("1.5");
+      const unstakeInsufficientAmount = this.minStake.mul(3);
```
Why is this changed?

R-Santev commented on test/helper.ts:
```diff
@@ -32,8 +32,8 @@ export async function initializeContext(context: any) {
   context.epochReward = hre.ethers.utils.parseEther("0.0000001");
-  context.minStake = hre.ethers.utils.parseEther("1");
-  context.minDelegation = hre.ethers.utils.parseEther("1");
+  context.minStake = hre.ethers.utils.parseEther("1000");
```
Why is this changed?

Vitomir2 commented:
@R-Santev some test files are modified because of tabulation and fixed small issues that caused tests to fail. Additionally, I have removed the `whitelisting` from the `registerValidator`, because it is a different purpose.

R-Santev commented on contracts/ValidatorSet/modules/Inspector/IInspector.sol:
```diff
@@ -0,0 +1,41 @@
+
+/**
+ * @notice Data type for the banned validators' withdrawals
+ * @param liquidTokens The amount of liquid tokens to be
```
description looks unfinished

R-Santev commented on contracts/ValidatorSet/modules/Inspector/Inspector.sol:
```diff
@@ -0,0 +1,98 @@
+
+        _withdraw(msg.sender, withdrawalBalance.withdrawableAmount);
+
+        delete withdrawalBalances[msg.sender];
```
reentrancy attack vector

Vitomir2 replied:
Quite a bad error. Thanks for this one and I will make sure to take note!

R-Santev commented on test/fixtures.ts:
```diff
@@ -239,7 +240,7 @@ async function vestingRewardsFixtureFunction(this: Mocha.Context) {
   );
 
-  await stakerValidatorSet.stake({ value: this.minStake });
+  await stakerValidatorSet.stake({ value: hre.ethers.utils.parseEther("1000") });
```
Why is this needed?

Vitomir2 replied:
Well, it is actually not needed as we discussed to just change the penalty and reporter reward in the Inspector contract.
